### PR TITLE
core,netty,testing: Support dup headers joined with commas

### PR DESCRIPTION
**Pull request by dapengzhang0**
Following the [spec](https://github.com/grpc/grpc/blob/master/doc/PROTOCOL-HTTP2.md) on duplicate header names:

**Custom-Metadata** header order is not guaranteed to be preserved except for values with duplicate header names. Duplicate header names may have their values joined with "," as the delimiter and be considered semantically equivalent. Implementations must split Binary-Headers on "," before decoding the Base64-encoded values.

### CODE REVIEW

File: netty/src/main/java/io/grpc/netty/GrpcHttp2HeadersUtils.java [hunk 1]
```diff
@@ -101,17 +101,34 @@ protected GrpcHttp2InboundHeaders newHeaders() {
...
+        if (!name.endsWith(binaryHeaderSuffix)) {
+          valueBytes = bytes(value);
+        } else {
+          indexOfComma = value.indexOf(',', startPos);
+          endPos = indexOfComma == AsciiString.INDEX_NOT_FOUND ? value.length() : indexOfComma;
+          curVal = value.subSequence(startPos, endPos, false);
+          valueBytes = BaseEncoding.base64().decode(curVal);
+          startPos = indexOfComma + 1;
+        }
+        if (namesAndValuesIdx == namesAndValues.length) {
```
Comment (ejona86) [inline netty/src/main/java/io/grpc/netty/GrpcHttp2HeadersUtils.java]: It's hard to see the flow now, especially for ascii values. Could the actual adding to the array be moved into a private method, like `add0(byte[] name, byte[] value, AsciiString valueStr)`? That would let you use the loop only for binary values.

File: testing/src/main/java/io/grpc/internal/testing/AbstractTransportTest.java [hunk 2]
```diff
@@ -790,14 +791,13 @@ public void basicStream() throws Exception {
...
+    serverHeaders.put(binaryKey, "dup,value");
     Metadata serverHeadersCopy = new Metadata();
     serverHeadersCopy.merge(serverHeaders);
     serverStream.writeHeaders(serverHeaders);
     Metadata headers = clientStreamListener.headers.get(TIMEOUT_MS, TimeUnit.MILLISECONDS);
     assertNotNull(headers);
-    assertEquals(
-        Lists.newArrayList(serverHeadersCopy.getAll(asciiKey)),
-        Lists.newArrayList(headers.getAll(asciiKey)));
+    assertAsciiMetadataValuesEqual(serverHeadersCopy.getAll(asciiKey),headers.getAll(asciiKey));
```
Comment (ejona86) [inline testing/src/main/java/io/grpc/internal/testing/AbstractTransportTest.java]: Add space after comma.

File: core/src/main/java/io/grpc/internal/TransportFrameUtil.java [hunk 3]
```diff
@@ -107,6 +119,35 @@
...
+        continue;
+      }
+      // Binary header
+      int prevIdx = 0;
+      for (int idx = 0; idx <= value.length; idx++) {
+        if (idx != value.length && value[idx] != (byte) ',') {
+          continue;
+        }
+        byte[] decodedVal =
+          BaseEncoding.base64().decode(new String(value, prevIdx, idx - prevIdx, US_ASCII));
```
Comment (ejona86) [inline core/src/main/java/io/grpc/internal/TransportFrameUtil.java]: Indent two more spaces.

**Comment by dapengzhang0** [comment]
PTALA

**Pull request merged**